src/libtls/tls_hkdf.c

```diff
@@ -0,0 +1,602 @@
+		chunk_t empty = chunk_empty;
+		chunk_t derived = chunk_from_str("tls13 derived");
+		if (!derive_secret(this, &derived, &empty))
+		{
+			DBG2(DBG_TLS, "unable to derive secret");
+			return FALSE;
+		}
+		return TRUE;
+	}
+
+	if (this->phase == HKDF_PHASE_1)
+	{
+		return TRUE;
+	}
+	else
+	{
+		DBG2(DBG_TLS, "invalid HKDF phase");
+		return FALSE;
+	}
+}
+
+static bool move_to_phase_2(private_tls_hkdf_t *this)
+{
+	if (this->phase == HKDF_PHASE_0)
+	{
+		move_to_phase_1(this);
+	}
+
+	if (this->phase == HKDF_PHASE_1)
+	{
+		this->salt = this->okm;
+
+		if (!this->shared_secret)
+		{
+			DBG2(DBG_TLS, "no shared secret set");
+			return FALSE;
+		}
+		else
+		{
+			memcpy(this->ikm.ptr, this->shared_secret->ptr, this->shared_secret->len);
+			this->ikm.len = this->shared_secret->len;
+		}
+
+		if (!extract(this))
+		{
+			DBG2(DBG_TLS, "unable extract PRK");
+			return FALSE;
+		}
+		this->phase = HKDF_PHASE_2;
+		/* prepare okm for next extract */
+		chunk_t empty = chunk_empty;
+		chunk_t derived = chunk_from_str("tls13 derived");
+		if (!derive_secret(this, &derived, &empty))
+		{
+			DBG2(DBG_TLS, "unable to derive secret");
+			return FALSE;
+		}
+		return TRUE;
+	}
+
+	if (this->phase == HKDF_PHASE_2)
+	{
+		return TRUE;
+	}
+	else
+	{
+		DBG2(DBG_TLS, "invalid HKDF phase");
+		return FALSE;
+	}
+}
+
+/*
+static bool move_to_phase_3(private_tls_hkdf_t *this)
+{
+	*/
+/*
+	 * (salt, ikm) -> prk
+	 *//*
+
+	if (this->phase == HKDF_PHASE_0)
+	{
+		move_to_phase_1(this);
+	}
+
+	if (this->phase == HKDF_PHASE_1)
+	{
+		move_to_phase_2(this);
+	}
+
+	if (this->phase == HKDF_PHASE_2)
+	{
+		this->salt = this->okm;
+
+		this->ikm = chunk_from_chars(
+			0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+			0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+			0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+			0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+		);
+
+		if (!extract(this))
+		{
+			DBG2(DBG_TLS, "unable extract PRK");
+			return FALSE;
+		}
+		this->phase = HKDF_PHASE_3;
+		return TRUE;
+	}
+
+	if (this->phase == HKDF_PHASE_3)
+	{
+		return TRUE;
+	}
+	else
+	{
+		DBG2(DBG_TLS, "invalid HKDF phase");
+		return FALSE;
+	}
+}
+*/
+
+static bool get_secret(private_tls_hkdf_t *this, chunk_t *secret)
+{
+	*secret = chunk_alloc(this->L);
+	memset(secret->ptr, 0, this->L);
+	memcpy(secret->ptr + secret->len - this->okm.len,
+		   this->okm.ptr, this->okm.len);
+	return TRUE;
+}
+
+METHOD(tls_hkdf_t, set_shared_secret, bool, private_tls_hkdf_t *this, chunk_t *shared_secret)
+{
+	if (shared_secret)
+	{
+		this->shared_secret = shared_secret;
+		return TRUE;
+	}
+	else
+	{
+		DBG2(DBG_TLS, "no shared secret provided");
+		return FALSE;
+	}
+}
+
+METHOD(tls_hkdf_t, generate_secret, bool, private_tls_hkdf_t *this, enum tls_hkdf_labels_t label, chunk_t *messages, chunk_t *secret)
+{
+	/*
+	 * 1. Prüfe Eingabeparameter
+	 * 2. Springe in Phase des gewünschten Secrets
+	 * 3. Gebe Secret zurück
+	 */
+	chunk_t label_name;
+
+	switch (label)
+	{
+		/*
```

Comment:
Maybe add an explanation to this commented out section why it is commented out - I guess it is because the PSK is not in scope (phase 1).